### PR TITLE
Graphics refactor #2

### PR DESCRIPTION
**Pull request by kkaefer**
- moves more code to the `gfx` namespace
- makes vertex descriptors `constexpr`

### CODE REVIEW

File: src/mbgl/gfx/index_buffer.hpp [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+protected:
+    IndexBufferResource() = default;
+public:
+    virtual ~IndexBufferResource() = default;
+};
+
+class IndexBuffer {
+public:
+    std::size_t elements;
+    std::unique_ptr<const IndexBufferResource> resource;
```
Comment (pozdnyakov) [inline src/mbgl/gfx/index_buffer.hpp]: Could we use `util::peer` instead (I guess it'd positively influence the binary size change)?

File: src/mbgl/gl/enum.cpp [hunk 2]
```diff
@@ -221,5 +221,25 @@ platform::GLenum Enum<gfx::CullFaceWindingType>::to(const gfx::CullFaceWindingTy
...
 }
 
+template <>
+gfx::BufferUsageType Enum<gfx::BufferUsageType>::from(const platform::GLint value) {
+    switch (value) {
+        case GL_STREAM_DRAW: return gfx::BufferUsageType::StreamDraw;
+        case GL_STATIC_DRAW: return gfx::BufferUsageType::StaticDraw;
+        case GL_DYNAMIC_DRAW: return gfx::BufferUsageType::DynamicDraw;
+    }
+    return {};
```
Comment (pozdnyakov) [inline src/mbgl/gl/enum.cpp]: assert(false) ?

File: src/mbgl/gfx/vertex_vector.hpp [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+#include <mbgl/util/ignore.hpp>
+
+#include <vector>
+
+namespace mbgl {
+namespace gfx {
+
+template <class V>
+class VertexVector {
```
[the remark this answers is not in the document]
Comment (kkaefer) [reply src/mbgl/gfx/vertex_vector.hpp]: I don't think there's any duplication in the generated code; all of the methods are almost certainly going to be inlined.

[the remark this answers is not in the document]
Comment (pozdnyakov) [reply src/mbgl/gfx/vertex_vector.hpp]: I mean just C++ code, not generated :-) they both provide the same semantics

**Comment by kkaefer** [comment]
> 40047f0

The fact that it's `constexpr` isn't the main benefit here; instead we're moving a lot of the validation of vertex types (i.e. that they're POD types and have a reasonable size) to static_asserts, as well as adding the information about the attributes and types contained in every vertex into the descriptors so that we can reference them later. Additionally, we've used enums that exactly matched the values that OpenGL defined; these are swapped out to be a lookup table.

**Pull request merged**